feat: scripts can now implement integrations

quintesse opened the pull request:
Needs #1630 to be applied first

quintesse committed:
feat: scripts can now implement integrations

Before this change JBang would only look for integrations
in the dependencies, now it also looks in the currently
compiled code.

quintesse commented:
So the discussion to merge this or not was about the issue that user files are now part of the class path when running the post build and if this could result in problems or not. Right now it only runs with all the dependencies specified by the user code, but not the code itself. (Which is why currently it isn't possible to write integration code within the user code).

A possible issue might be that the classloader would hold a lock on the very resources that the integration might try to change, something that especially on Windows might cause problems.

maxandersen commented:
with #1641 this should be doable without "polluting" the integration class path. 

But it does not yet work because the integration assumes all deps have a GAV...

maxandersen committed:
test: added integration tests for in-script integrations

Tests use --fresh to ensure integrations are re-executed and not
served from cache, consistent with existing integration tests.

maxandersen commented:
The GAV issue mentioned [here](https://github.com/jbangdev/jbang/pull/1634#issuecomment-1626899080) is no longer relevant — `IntegrationManager` now handles deps without a GAV by falling back to the filename as key:

```java
if (art.getCoordinate() != null) { // skipping dependencies that does not have a GAV
    deps.put(art.getCoordinate().toCanonicalForm(), art.getFile());
} else {
    deps.put(art.getFile().getFileName().toString(), art.getFile());
}
```

Also cherry-picked and rebased onto current main. Tests now pass (needed `--fresh` to avoid cached builds skipping integration execution). Branch: `pr-1634-cherry`

maxandersen commented:
**Status update:** Cherry-picked onto current main and pushed to `pr-1634-cherry`. All 6 integration tests pass (including 4 new in-script integration tests).

**Summary of changes:**
- `IntegrationManager.getClassLoader()` now includes the compile output dir, so scripts can implement their own integrations via `META-INF/jbang-integration.list`
- Keeps main's improvements: try-with-resources for the classloader, `@NonNull` (jspecify)
- The prerequisite PRs (#1630, #1641) are already merged

**Previously raised concerns — all resolved:**
- **GAV issue** (maxandersen, July 2023): was already fixed in May 2022 via #1350 — deps without a GAV fall back to filename as key
- **Test failures** (quintesse, June 2025): caused by missing `--fresh` flag — cached builds skip integration execution entirely. Fixed by adding `--fresh` to all 4 new tests, consistent with the existing integration tests in the same file

Ready for review as a fresh PR from `pr-1634-cherry`, or this PR can be retargeted to that branch.
